LeaderboardBackend/Services/Impl/UserService.cs

```diff
@@ -21,23 +20,50 @@ public UserService(ApplicationContext applicationContext)
 
     public async Task<User?> GetUserByEmail(string email)
     {
-        return await _applicationContext.Users.FirstOrDefaultAsync(user => user.Email == email);
+        return await _applicationContext.Users.SingleOrDefaultAsync(user => user.Email == email);
     }
 
     public async Task<User?> GetUserByName(string name)
     {
-        string lowerName = name.ToLower();
-
-        // We save a username with casing, but match without.
-        // Effectively you can't have two separate users named e.g. "cool" and "cOoL".
-        return await _applicationContext.Users.FirstOrDefaultAsync(
-            user => user.Username != null && user.Username.ToLower() == lowerName
-        );
+        return await _applicationContext.Users.SingleOrDefaultAsync(user => user.Username == name);
     }
 
-    public async Task CreateUser(User user)
+    public async Task<CreateUserResult> CreateUser(RegisterRequest request)
     {
-        _applicationContext.Users.Add(user);
+        /* Linq query translated as:
+        SELECT count(*) FILTER (WHERE t.username = @__request_Username_0)::int > 0 AS "usernameExists",
+            count(*) FILTER (WHERE t.email = @__request_Email_1)::int > 0 AS "emailExists"
+        FROM (
+            SELECT u.email, u.username, 1 AS "Key"
+            FROM users AS u
+        ) AS t
+        GROUP BY t."Key"
+        LIMIT 2
+        */
+        var conflicts = await _applicationContext.Users.GroupBy(x => 1)
+            .Select(users => new
+            {
+                usernameExists = users.Count(x => x.Username == request.Username) > 0,
+                emailExists = users.Count(x => x.Email == request.Email) > 0
+            }).SingleAsync();
+
+        if (conflicts.usernameExists || conflicts.emailExists)
+        {
+            return new CreateUserConflicts(Username: conflicts.usernameExists, Email: conflicts.emailExists);
+        }
```

Comment:
The alternative is catching an exception when inserting and checking which constraint was violated, but I think only one violated constraint at a time will be reported.

Reply:
That's the proper way to do it. Catch the exception and then return the corresponding error code.

Reply:
Yeah you're right, because someone could register with the same username in between the check and the insertion.

With other database providers you would have to parse the error message to get the constraint name, but with Postgres' provider it seems the violated constraint name is a property on the exception.

Reply:
changed